PID updates to include additional filters

lthall opened the pull request:
This is the first step in a number of PID object updates to address various control challenges I have been seeing with aircraft of various sizes. These changes do not impact the performance if the parameter is left at zero.

I have also found a couple of alt hold initialisation bugs that needed to be fixed to test this properly.

This update adds filtering to address issues with noise entering the controllers through the unfiltered EKF and noise caused by the EKF itself and things like lower sample rate RC inputs.

These updates also support feed forward natively and will address control issues with aircraft controlled by aerodynamic surfaces like single and coax copter.

Heli pids have been simplified but can be simplified further since the internal target filter can be used instead of the FF filter currently used in Heli. This will need to be something the Heli maintainers look at and decide.

I am not sure if I have correctly handled the antenna tracker and sail boat changes.

lthall commented:
For some reason the first test on travis didn't complete but it did on my branch. I have found that I need to hit restart from time to time but I can't do it here.

lthall commented on libraries/AC_PID/AC_HELI_PID.cpp:
```diff
@@ -68,13 +68,6 @@ float AC_HELI_PID::get_leaky_i(float leak_rate)
             _integrator -= (float)(_integrator + _leak_min) * leak_rate;
         }
 
-        _integrator += ((float)_input * _ki) * _dt;
```
The integration and limiting is all done in the existing pid object now.

We also have FF so this part of the PID object could replace the FF in the Heli code too.

lthall commented:
I also changed the .param file names to update _FILT to _FLTD or _FLTE.

lthall commented:
I also noticed that a couple other vehicles were using the input filter option. Fixed that too.

khancyr commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp:
```diff
@@ -283,56 +283,27 @@ void AC_AttitudeControl_Heli::update_althold_lean_angle_max(float throttle_in)
-    float roll_pd, roll_i, roll_ff;             // used to capture pid values
-    float pitch_pd, pitch_i, pitch_ff;          // used to capture pid values
-    float rate_roll_error_rads, rate_pitch_error_rads;    // simply target_rate - current_rate
+    float roll_pid, roll_ff;             // used to capture pid values
```
it is better to directly assign value on object creation in cpp as it will use the object constructor to assign the value instead of doing it in two steps : varaible create then assignement !

lthall replied:
I am maintaining consistency with the way Heli has been done here already. I would be happy to work with Bill and Chris to do more though.

khancyr commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp:
```diff
@@ -359,57 +330,33 @@ void AC_AttitudeControl_Heli::rate_bf_to_motor_roll_pitch(const Vector3f &rate_r
 {
-    float pd,i,vff;     // used to capture pid values for logging
-    float rate_error_rads;       // simply target_rate - current_rate
+    float pid, vff;     // used to capture pid values for logging
```
same as previous comment

lthall replied:
Same as previous

khancyr commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
-            _integrator = -_imax;
-        } else if (_integrator > _imax) {
-            _integrator = _imax;
+    if(!is_zero(_ki) && is_positive(_dt)) {
```
if (

lthall replied:
Good point, I will do a full formatting check.

khancyr commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
 {
-    return get_p() + get_i();
+    _pid_info.FF = _target * _kff;
+    return _target * _kff;
```
better return _pid_info.FF as you already made the calculus

lthall replied:
I discussed this with Randy and as we don't control this object we are not comfortable using it as the primary storage medium. I don't know if this is the right decision or not.

lthall replied:
All discussion welcome here!!

khancyr replied:
sorry, I don't understand where we are not controling this object ? _pid_info is part of AC_PID and we aren't doing multthreading... so it is safe. Otherwise we get a bigger issue and need protect with mutex ! 
my point was to prevent to do the calculus twice. I am even confident that the compiler will create a temp variable to do that on its own.

lthall replied:
My basic concern here is the pid object is accessing the logging object to get it's data rather than the other way round. I am sure you are correct that this isn't going to cause an issue in it's current form I can't see this being the right way to do it. Someone making changes to the logging behaviour could mess with the control of aircraft.

You are also right that I am doing the same calculation twice so I could remove that without depending on the pid_info structure.

khancyr commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
 {
-    return get_p() + get_i() + get_d();
+    _pid_info.FF = target * _kff;
+    return target * _kff;
```
better return _pid_info.FF as you already made the calculus

lthall replied:
same as above.

khancyr commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
-    // derivative component
-    _pid_info.D = (_kd * _derivative);
-    return _pid_info.D;
+    return _error * _kp;
```
can't we just get _pid_info.P ?

lthall replied:
same as above.

khancyr commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
 
+float AC_PID::get_d() const
+{
+    return _kd * _derivative;
```
can't we just get _pid_info.D ?

lthall replied:
same as above.

lthall commented:
I have precalculated the values going into the _pid_info so it isn't calculated twice.

lthall commented on libraries/AP_Motors/AP_Motors_Class.h:
```diff
@@ -82,8 +82,11 @@ class AP_Motors {
 
     // set_roll, set_pitch, set_yaw, set_throttle
     void                set_roll(float roll_in) { _roll_in = roll_in; };        // range -1 ~ +1
+    void                set_roll_ff(float yaw_in) { _roll_in_ff = yaw_in; };    // range -1 ~ +1
```
Thanks!

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -8,45 +8,57 @@ const AP_Param::GroupInfo AC_PID::var_info[] = {
-    AP_GROUPINFO("FF",   7, AC_PID, _ff, 0),
+    AP_GROUPINFO("FF", 7, AC_PID, _kff, 0),
+
+    // @Param: FILT
```
Done and picked up a couple other comment changes

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
+}
 
-    // sanity check _filt_hz
-    _filt_hz = MAX(_filt_hz, AC_PID_FILT_HZ_MIN);
```
Yes, this is to enable the filters to be disabled by setting the frequency to zero.

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
+    } else {
+        float error_last = _error;
+        _target += get_filt_T_alpha() * (target - _target);
+        _error += get_filt_E_alpha() * ((_target - measurement)-_error);
```
Done

lthall commented on ArduCopter/tuning.cpp:
```diff
@@ -215,7 +215,7 @@ void Copter::tuning()
 
      case TUNING_RATE_YAW_FILT:
-         attitude_control->get_rate_yaw_pid().filt_hz(tuning_value);
+         attitude_control->get_rate_yaw_pid().filt_D_hz(tuning_value);
```
Nice Catch!

lthall commented on ArduCopter/system.cpp:
```diff
@@ -595,7 +595,7 @@ void Copter::allocate_motors(void)
         break;
     case AP_Motors::MOTOR_FRAME_TRI:
-        attitude_control->get_rate_yaw_pid().filt_hz().set_default(100);
+        attitude_control->get_rate_yaw_pid().filt_D_hz().set_default(100);
```
This is one of the areas where the filter settings should change.

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:
```diff
@@ -1110,7 +1041,7 @@ float AC_AttitudeControl::max_rate_step_bf_pitch()
 float AC_AttitudeControl::max_rate_step_bf_yaw()
 {
-    float alpha = get_rate_yaw_pid().get_filt_alpha();
+    float alpha = get_rate_yaw_pid().get_filt_D_alpha();
```
Good catch, to be generic this should be the minimum of the E and D filters.

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp:
```diff
@@ -283,56 +283,27 @@ void AC_AttitudeControl_Heli::update_althold_lean_angle_max(float throttle_in)
-		}else{
-			pitch_i = _pid_rate_pitch.get_i();
-		}
+    if (_flags_heli.leaky_i){
```
I did a general format check on the file.

lthall commented:
> - when doing a commit affecting multiple libraries/vehicles we use _Global: _ as the prefix in the commit message (instead of _ArduPilot: _)
I changed the commit prefix.

> - I would have separated the adding of error since that's new functionality
> - I think that change in the commit message should be rename (otherwise someone might think there is an actual functionality change there when it's just a rename)
> - I've commented below on a number of lines where a rename happens in a later commit; it would be better if those happened in that first commit (after editing the first commit to make the same renames, it will disappear from later commits so it isn't very hard to change)

With my knowledge of GIT this is very difficult for me to do. Where I can I do try to do this but there is about three months of work here where working out what the names are is the last step. It would be great to learn how to insert a simple change without breaking all my work after that. (Yeh I know for many people I am looking incompetent right now, I am :) )

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl.h:
```diff
@@ -321,7 +321,7 @@ class AC_AttitudeControl {
 
     // Run the yaw angular velocity PID controller and return the output
-    virtual float rate_target_to_motor_yaw(float rate_actual_rads, float rate_target_rads);
+    virtual float rate_target_to_motor_yaw(float rate_actual_rads, float rate_target_rads) { return 0;}
```
I will need some help to get this right.

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
+        // calculate and filter derivative
+        if (_dt > 0.0f) {
+            float derivative = (_error - error_last) / _dt;
+            _derivative = _derivative + get_filt_D_alpha() * (derivative - _derivative);
```
Done

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
+        }
     }
 
     // update filter and calculate derivative
```
Yep.

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
+        // calculate and filter derivative
+        if (_dt > 0.0f) {
+            float derivative = (_error - error_last) / _dt;
+            _derivative = _derivative + get_filt_D_alpha() * (derivative - _derivative);
```
Done

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -73,108 +87,149 @@ void AC_PID::set_dt(float dt)
+float AC_PID::get_ff()
 {
-    return get_p() + get_i();
+    _pid_info.FF = _target * _kff;
```
It is used separately so no but I too would like to think of a more elegant way of doing this. I don't have one yet. I think it would involve providing a pointer to the PID object into the motors libary rather than just the PID and FF terms.

lthall commented on libraries/AC_PID/AC_HELI_PID.h:
```diff
@@ -17,7 +17,7 @@ class AC_HELI_PID : public AC_PID {
 
     /// Constructor for PID
-    AC_HELI_PID(float initial_p, float initial_i, float initial_d, float initial_imax, float initial_filt_hz, float dt, float initial_ff);
+    AC_HELI_PID(float initial_p, float initial_i, float initial_d, float initial_imax, float initial_filt_T_hz, float initial_filt_E_hz, float initial_filt_D_hz, float dt, float initial_ff);
```
Legend!

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp:
```diff
@@ -283,56 +283,27 @@ void AC_AttitudeControl_Heli::update_althold_lean_angle_max(float throttle_in)
-		}else{
-			roll_i = _pid_rate_roll.get_i();
-		}
+    if (_flags_heli.leaky_i){
```
Done

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.h:
```diff
@@ -36,9 +36,10 @@ class AC_AttitudeControl_Heli : public AC_AttitudeControl {
-        _pid_rate_roll(AC_ATC_HELI_RATE_RP_P, AC_ATC_HELI_RATE_RP_I, AC_ATC_HELI_RATE_RP_D, AC_ATC_HELI_RATE_RP_IMAX, AC_ATC_HELI_RATE_RP_FILT_HZ, dt, AC_ATC_HELI_RATE_RP_FF),
-        _pid_rate_pitch(AC_ATC_HELI_RATE_RP_P, AC_ATC_HELI_RATE_RP_I, AC_ATC_HELI_RATE_RP_D, AC_ATC_HELI_RATE_RP_IMAX, AC_ATC_HELI_RATE_RP_FILT_HZ, dt, AC_ATC_HELI_RATE_RP_FF),
-        _pid_rate_yaw(AC_ATC_HELI_RATE_YAW_P, AC_ATC_HELI_RATE_YAW_I, AC_ATC_HELI_RATE_YAW_D, AC_ATC_HELI_RATE_YAW_IMAX, AC_ATC_HELI_RATE_YAW_FILT_HZ, dt, AC_ATC_HELI_RATE_YAW_FF),
+        _passthrough_roll(0), _passthrough_pitch(0), _passthrough_yaw(0),
```
I might have accidentally pulled them in when I rebased off the old code.

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.h:
```diff
@@ -36,9 +36,10 @@ class AC_AttitudeControl_Heli : public AC_AttitudeControl {
+        _passthrough_roll(0), _passthrough_pitch(0), _passthrough_yaw(0),
+        _pid_rate_roll(AC_ATC_HELI_RATE_RP_P, AC_ATC_HELI_RATE_RP_I, AC_ATC_HELI_RATE_RP_D, AC_ATC_HELI_RATE_RP_IMAX, 0.0f, AC_ATC_HELI_RATE_RP_FILT_HZ, 0.0f, dt, AC_ATC_HELI_RATE_RP_FF),
+        _pid_rate_pitch(AC_ATC_HELI_RATE_RP_P, AC_ATC_HELI_RATE_RP_I, AC_ATC_HELI_RATE_RP_D, AC_ATC_HELI_RATE_RP_IMAX, 0.0f, AC_ATC_HELI_RATE_RP_FILT_HZ, 0.0f, dt, AC_ATC_HELI_RATE_RP_FF),
+        _pid_rate_yaw(AC_ATC_HELI_RATE_YAW_P, AC_ATC_HELI_RATE_YAW_I, AC_ATC_HELI_RATE_YAW_D, AC_ATC_HELI_RATE_YAW_IMAX, 0.0f, AC_ATC_HELI_RATE_YAW_FILT_HZ, 0.0f, dt, AC_ATC_HELI_RATE_YAW_FF),
```
Yep, that one I should have got!!!

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -581,48 +582,25 @@ void AC_PosControl::run_z_controller()
     if (_flags.reset_accel_to_throttle) {
-        // Reset Filter
-        _accel_error.z = 0;
+        _pid_accel_z.set_integrator((_attitude_control.get_throttle_in()-_motors.get_throttle_hover())*1000.0f);
```
Done

lthall commented on libraries/APM_Control/AR_AttitudeControl.cpp:
```diff
@@ -354,36 +354,13 @@ float AR_AttitudeControl::get_steering_out_rate(float desired_rate, bool motor_l
-    // get d
-    const float d = _steer_rate_pid.get_d();
-
+    float output = _steer_rate_pid.update_all(_desired_turn_rate, _ahrs.get_yaw_rate_earth(), (motor_limit_left||motor_limit_right));
```
Done

lthall commented on libraries/APM_Control/AR_AttitudeControl.cpp:
```diff
@@ -613,7 +548,7 @@ float AR_AttitudeControl::get_desired_pitch() const
     }
 
-    return _pitch_to_throttle_pid.get_pid_info().desired;
+    return _pitch_to_throttle_pid.get_pid_info().target;
```
Ok, Git retard alert.

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -5,61 +5,60 @@
+// @Units: Hz
+AP_GROUPINFO("FLTE", 9, AC_PID, _filt_E_hz, AC_PID_EFILT_HZ_DEFAULT),
+
+AP_GROUPEND };
```
Yep, I must have missed this when I removed the auto formatting changes.

lthall replied:
Fixed

lthall commented on AntennaTracker/servos.cpp:
```diff
@@ -197,8 +195,7 @@ void Tracker::update_yaw_position_servo()
 
-    g.pidYaw2Srv.set_input_filter_all(nav_status.angle_error_yaw);
-    float servo_change = g.pidYaw2Srv.get_pid();
+    float servo_change = g.pidYaw2Srv.update_error(nav_status.angle_error_yaw);;
```
Done

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -198,30 +262,63 @@ void AC_PID::save_gains()
+}
+
+// calc_filt_alpha - recalculate the input filter alpha
+float AC_PID::get_filt_D_alpha() const
```
Done

lthall commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -198,30 +262,63 @@ void AC_PID::save_gains()
+    return get_filt_alpha(_filt_D_hz);
+}
+
+// calc_filt_alpha - recalculate the input filter alpha
```
done

lthall commented on libraries/AP_WheelEncoder/AP_WheelRateControl.h:
```diff
@@ -58,8 +58,8 @@ class AP_WheelRateControl {
-    AC_PID          _rate_pid0 = AC_PID(AP_WHEEL_RATE_CONTROL_P, AP_WHEEL_RATE_CONTROL_I, AP_WHEEL_RATE_CONTROL_D, AP_WHEEL_RATE_CONTROL_IMAX, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_DT, AP_WHEEL_RATE_CONTROL_FF);
-    AC_PID          _rate_pid1 = AC_PID(AP_WHEEL_RATE_CONTROL_P, AP_WHEEL_RATE_CONTROL_I, AP_WHEEL_RATE_CONTROL_D, AP_WHEEL_RATE_CONTROL_IMAX, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_DT, AP_WHEEL_RATE_CONTROL_FF);
+    AC_PID          _rate_pid0 = AC_PID(AP_WHEEL_RATE_CONTROL_P, AP_WHEEL_RATE_CONTROL_I, AP_WHEEL_RATE_CONTROL_D, AP_WHEEL_RATE_CONTROL_FF, AP_WHEEL_RATE_CONTROL_IMAX, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_DT);
+    AC_PID          _rate_pid1 = AC_PID(AP_WHEEL_RATE_CONTROL_P, AP_WHEEL_RATE_CONTROL_I, AP_WHEEL_RATE_CONTROL_D, AP_WHEEL_RATE_CONTROL_FF, AP_WHEEL_RATE_CONTROL_IMAX, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_FILT, AP_WHEEL_RATE_CONTROL_DT);
```
All zero by default.

lthall commented on libraries/AC_PID/AC_PID.h:
```diff
@@ -9,102 +9,129 @@
+    float get_filt_T_alpha() const;
+    float get_filt_E_alpha() const;
+    float get_filt_D_alpha() const;
+    float get_integrator() const { return _integrator; }
```
true. Removed

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -379,7 +379,8 @@ void AC_PosControl::relax_alt_hold_controllers(float throttle_setting)
     _accel_last_z_cms = 0.0f;
     _accel_target.z = -(_ahrs.get_accel_ef_blended().z + GRAVITY_MSS) * 100.0f;
-    _flags.reset_accel_to_throttle = true;
+    _flags.reset_rate_to_accel_z = true;
```
yes, this is an existing bug I found during testing. It should go in as a separate patch probably but it was here already.

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Sub.cpp:
```diff
@@ -169,9 +169,9 @@ const AP_Param::GroupInfo AC_AttitudeControl_Sub::var_info[] = {
-	_pid_rate_yaw(AC_ATC_SUB_RATE_YAW_P, AC_ATC_SUB_RATE_YAW_I, AC_ATC_SUB_RATE_YAW_D, AC_ATC_SUB_RATE_YAW_IMAX, AC_ATC_SUB_RATE_YAW_FILT_HZ, dt)
+    _pid_rate_roll(AC_ATC_SUB_RATE_RP_P, AC_ATC_SUB_RATE_RP_I, AC_ATC_SUB_RATE_RP_D, 0.0f, AC_ATC_SUB_RATE_RP_IMAX, AC_ATC_SUB_RATE_RP_FILT_HZ, 0.0f, AC_ATC_SUB_RATE_RP_FILT_HZ, dt),
+    _pid_rate_pitch(AC_ATC_SUB_RATE_RP_P, AC_ATC_SUB_RATE_RP_I, AC_ATC_SUB_RATE_RP_D, 0.0f, AC_ATC_SUB_RATE_RP_IMAX, AC_ATC_SUB_RATE_RP_FILT_HZ, 0.0f, AC_ATC_SUB_RATE_RP_FILT_HZ, dt),
+	_pid_rate_yaw(AC_ATC_SUB_RATE_YAW_P, AC_ATC_SUB_RATE_YAW_I, AC_ATC_SUB_RATE_YAW_D, 0.0f, AC_ATC_SUB_RATE_YAW_IMAX, AC_ATC_SUB_RATE_YAW_FILT_HZ, 0.0f, AC_ATC_SUB_RATE_YAW_FILT_HZ, dt)
```
I removed the target filter but pretty much everybody should start using them.
Fixed.

lthall commented:
Ok after much GIT mashing I have addressed the changes above and squashed them into the previous commits to minimise the number of commits.

I also separated out the Alt_Hold commit as a separate commit and made it first.

Francisco will do some magic on this in the morning to address one of his comments for me.

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp:
```diff
@@ -359,57 +330,33 @@ void AC_AttitudeControl_Heli::rate_bf_to_motor_roll_pitch(const Vector3f &rate_r
-        if (((AP_MotorsHeli&)_motors).rotor_runup_complete()) {
-            i = _pid_rate_yaw.get_i();
-        } else {
-            i = ((AC_HELI_PID&)_pid_rate_yaw).get_leaky_i(AC_ATTITUDE_HELI_RATE_INTEGRATOR_LEAK_RATE);    // If motor is not running use leaky I-term to avoid excessive build-up
```
Yes. This is part of the standard PID now.

bnsgeyer commented:
@lthall this completes heli changes to bring it in line with your PID changes.  Hopefully I caught all of the suggested changes to make the code more efficient.

lthall commented:
Thanks Bill!

I think we are pretty much ready to pull the trigger here!!

OXINARF commented:
I've worked with @lthall to fix some stuff (I've also fixed an issue with @bnsgeyer's code - his changes were based on an older code version so it carried an issue that Leonard had already fixed).

That's now pushed on this PR - everything was squashed up (but if you want to see the fixes: https://github.com/OXINARF/ardupilot/commit/ced1d6863634f3fec9890a41f6f8119062bceed4, https://github.com/OXINARF/ardupilot/commit/1d1820b245d199b8fd3847ad425d3a48eacc5198, https://github.com/OXINARF/ardupilot/commit/0d71af30a786fe30442299a337b6fc0fc11e1b88, https://github.com/OXINARF/ardupilot/commit/634ab55f41155ddb367bf7c3b462d4f91ad584e2, https://github.com/OXINARF/ardupilot/commit/c2ae4400e1ae76fd7878e160fefca83810640f0a, https://github.com/OXINARF/ardupilot/commit/28cd5ed5af2f47870d851aa06e02d63996632356).


I think this is now ready to go. Unless there is an objection in the next day or so, I'll merge it.

rmackay9 commented:
It looks like AP_Motor's _roll_in_ff isn't used anywhere.  Normally we don't add code that isn't used.

rmackay9 commented on libraries/AC_PID/AC_PID.h:
```diff
@@ -9,102 +9,128 @@
+    AP_Float &kI() { return _ki; }
+    AP_Float &kD() { return _kd; }
+    AP_Float &ff() { return _kff;}
+    AP_Float &filt_T_hz() { return _filt_T_hz; }
```
no a blocker for this PR but we don't actually use these new accessors so they could be removed.  It might be good to be sure that we don't need to set the defaults in Copter's allocate_motors() function which does seem to set some defaults for kP, kI and dD.

rmackay9 commented on libraries/AC_PID/AC_PID.h:
```diff
@@ -9,102 +9,128 @@
     // set the desired and actual rates (for logging purposes)
-    void        set_desired_rate(float desired) { _pid_info.desired = desired; }
-    void        set_actual_rate(float actual) { _pid_info.actual = actual; }
+    void set_desired_rate(float desired) { _pid_info.target = desired; }
```
we seem to be some mixing of "desired" and "target".. not sure if we can consolidate that or not.  Not a blocker but just saying we've introduced a difference with this PR by renaming "desired" to "target" in the PID info but not in other places.

lthall replied:
done

rmackay9 commented on libraries/AC_PID/AC_HELI_PID.cpp:
```diff
@@ -23,17 +23,17 @@ const AP_Param::GroupInfo AC_HELI_PID::var_info[] = {
     // @DisplayName: Velocity FF FeedForward Gain
     // @Description: Velocity FF Gain which produces an output value that is proportional to the demanded input
-    AP_GROUPINFO("VFF",    4, AC_HELI_PID, _ff, 0),
+    AP_GROUPINFO("VFF",    4, AC_HELI_PID, _kff, 0),
```
should we perhaps name the parameter variable "_vff" instead of "_kff" to make it match the parameter name?

lthall replied:
addressed with parameter ordering

rmackay9 commented on libraries/AC_PID/AC_HELI_PID.cpp:
```diff
@@ -23,17 +23,17 @@ const AP_Param::GroupInfo AC_HELI_PID::var_info[] = {
     // @DisplayName: PID Integral Maximum
     // @Description: The maximum/minimum value that the I term can output
-    AP_GROUPINFO("IMAX", 5, AC_HELI_PID, _imax, 0),
+    AP_GROUPINFO("IMAX", 5, AC_HELI_PID, _kimax, 0),
```
any reason for the rename of _imax to _kimax?

rmackay9 commented on libraries/AC_PID/AC_HELI_PID.cpp:
```diff
@@ -48,16 +48,16 @@ const AP_Param::GroupInfo AC_HELI_PID::var_info[] = {
 // if the error is zero.
 
-float AC_HELI_PID::get_leaky_i(float leak_rate)
+void AC_HELI_PID::get_leaky_i(float leak_rate)
```
the function name is a bit weird.  normally a function starting with "get_" will return something but here it doesn't.

lthall replied:
Changed to update_leaky_i

rmackay9 commented on libraries/AP_WheelEncoder/AP_WheelRateControl.cpp:
```diff
@@ -144,31 +144,9 @@ float AP_WheelRateControl::get_rate_controlled_throttle(uint8_t instance, float
-
     // constrain and set limit flags
-    float output = ff + p + i + d;
+    float output = rate_pid.update_all(desired_rate, actual_rate, (_limit[instance].lower || _limit[instance].upper));
```
There's a slight change in logic here with the limiting flags.  I think it will cause the incorrect behaviour in some (perhaps unlikely) cases.  For example if FF, P and D are zero so that all the output is coming from I, the output will get stuck the moment the output hits a limit.

lthall replied:
added missing ff terms

rmackay9 commented on libraries/AC_PID/AC_PID.cpp:
```diff
@@ -72,109 +86,157 @@ void AC_PID::set_dt(float dt)
+//  target and error are filtered
+//  the derivative is then calculated and filtered
+//  the integral is then updated based on the setting of the limit flag
+float AC_PID::update_all(float target, float measurement, bool limit)
```
the name is a bit unintuitive perhaps.  It's an "update" function but it actually returns a value.. but that value isn't really "all" I think because it's missing the FF?

rmackay9 commented on ArduPlane/quadplane.cpp:
```diff
@@ -495,7 +495,7 @@ const AP_Param::ConversionInfo q_conversion_table[] = {
     { Parameters::k_param_quadplane, 80,   AP_PARAM_FLOAT, "Q_P_ACCZ_I"},      //  Q_AZ_I
     { Parameters::k_param_quadplane, 144,  AP_PARAM_FLOAT, "Q_P_ACCZ_D"},      //  Q_AZ_D
     { Parameters::k_param_quadplane, 336,  AP_PARAM_FLOAT, "Q_P_ACCZ_IMAX"},   //  Q_AZ_IMAX
-    { Parameters::k_param_quadplane, 400,  AP_PARAM_FLOAT, "Q_P_ACCZ_FILT"},   //  Q_AZ_FILT
```
we should check that @tridge is happy with this.  I think it is OK because those upgrading from a very old firmware (which used Q_AZ_FILT) will get the value copied over to the new parameter name (Q_P_ACCZ_FILTD) while those who were using a new firmware (Q_P_ACCZ_FILT) will be unaffected because the parameter is not moved in eeprom but rather just renamed.

I think the only people that will be caught out is those who have a stored parameter file somewhere.

rmackay9 commented:
In general it looks OK.  I think we should test Rover and Tracker in SITL at least to be sure they are working OK.

rmackay9 commented:
I'm planning to add some parameter conversion to this PR, hopefully tomorrow.

rmackay9 commented:
I've created 3 commits which can be pulled on top of this PR here: https://github.com/rmackay9/rmackay9-ardupilot/commits/lthall-filt-param

These fixup the parameter conversion code but the quadplane conversion doesn't work for some reason that I do not understand

rmackay9 commented:
Just testing Param conversions more completely and it seems like Sub's ATC_RAT_PIT_FF, RLL_FF and YAW_FF are being set to zero..

rmackay9 commented:
I've got a [new commit in this branch](https://github.com/rmackay9/rmackay9-ardupilot/commits/lthall-filt-param) which resolves the Copter's FF parameter conversion.

Sadly we had the same issue for a bunch of Rover PID objects so we need some more parameter conversions I'm afraid.

rmackay9 commented:
OK, this is passing travis/ci now.  So we just need to ensure that all the affected vehicles are still ok.

I'll test Rover..
@EbinPhilip could you test BalanceBot?
@IamPete1, could you test sailboat and/or antenna tracker?
@tridge, @WickedShell, do you want to test quadplane or should we try and test and provide logs?

EbinPhilip commented:
Sure. I'll test it this weekend.

IamPete1 commented:
Just had a go on tracker, for me this caused bad P oscillations, I got no filtering on after a flash and had to manual set FILTE to 0.1 then it was OK in pitch, but had to also half YAW_P. not sure if it due to not having the param conversion or a error in the defaults. Once the params were changed it works great!

Sailboat changes look OK but not tested, hopefully will be able to soonish.

lthall commented:
Just something to consider for people working on the various vehicles. The FLTT is the filter that will remove noise that comes in through the EKF that is not filtered by any of the other filters. You will find less noise on your control signals. For most situations you will benifit by setting FLTT to be the same as your FLTT = FLTD or FLTE

tridge commented:
needs testing in realflight for quadplanes

kd0aij commented:
SITL/RealFlight8 testing:

AddictionX tailsitter: **OK** with no changes to params
Convergence (twin-engine tilt-vectored tricopter): **OK** with no changes to params

FT_Bronco_Pete (twin-engine TVBS): **OK** after QAUTOTUNE (copter-frame yaw oscillations before tuning) 

Stryker_quad (copter tailsitter): pitch control is **broken** in QHOVER with no changes to params
(insufficient pitch authority; doesn't hold vertical orientation)
log: https://drive.google.com/open?id=1btTasxiW9VpJ0Liw4_4LRSlmyuhbWpdL

rmackay9 commented:
@kd0aij, thanks very much for testing.  Re the Stryker_quad, I guess that master is working OK so it's pretty clear that it's this PR that has broken things?

lthall commented:
![image](https://user-images.githubusercontent.com/100896/61753659-536dbb00-adef-11e9-9bb2-416877853995.png)

I don't think this is a problem with the PID's. The output is moving past the limits and still getting little response. I think there is a problem with this frame. We have not changed anything in plane so I can't see how this is our changes.

lthall commented:
@kd0aij I have went over this log. From what I can see of the videos of this aircraft and the logs this should be a standard quad with yaw enhancement from the ailerons. The outputs don't look like quad at all. 

Considering that the same quad works in both SITL and I have tested the X frame on a number of quads I suspect something is happening in Plane motor mixing with plane throttle.

This is just Master with no changes right?

It would be great if you could provide the equivalent flight test with Master.

kd0aij commented:
Stryker_quad (copter tailsitter) on latest master (1fbd41e6) is OK:
https://drive.google.com/open?id=1E2yNMP9p_tTnUjpsQMi13s3ig-NRPr8F

tridge commented:
@kd0aij Leonard and I have found what we think the issue was with the stryker. It now flies for me about the same as master. Can you please re-test?

lthall commented:
@kd0aij Sorry about messing you around with that one. I didn't realise you weren't using the same motor mixer as copter there.

You may need to set roll and pitch FF to zero as the FF was not getting passed in before.

kd0aij commented:
@lthall  @tridge  Thanks, ef1cb8e (pretty obviously) fixes the problem with copter tailsitters, and I agree that the Stryker_quad now looks OK in SITL.

I'll rebase https://github.com/ArduPilot/ardupilot/pull/11195 on this and test to make sure it doesn't adversely affect @losawing

lthall commented:
@kd0aij Thanks!!

kd0aij commented:
@lthall I couldn't find any code changes which would have affected the behavior of the tailsitter "bodyframe roll" options, and tested in RF8 again... now it seems OK.  Sorry about the confusion, I must have made a mistake with a parameter value.

I'll rebase my gainscaling PR on this and test again, but this PR looks good to me for tailsitters.

rmackay9 commented:
I've tested with rover (both ackerman and skid-steering) and the outputs for turn rate and speed are identical.

rmackay9 commented:
travis seems confused because it's reporting that travis is "pending" but it's actually completed successfully.

rmackay9 commented:
I'm going to merge this within a couple of hours unless anybody would like me to delay the merge..

rmackay9 commented:
Merged, thanks!

amilcarlucas commented on libraries/AP_Logger/LogStructure.h:
```diff
@@ -1224,10 +1225,10 @@ struct PACKED log_Arm_Disarm {
-#define PID_FMT    "Qffffff"
-#define PID_UNITS  "s------"
-#define PID_MULTS  "F------"
+#define PID_LABELS "TimeUS,Des,Act,Err,P,I,D,FF"
```
missing s/Des/Tar/ ?

lthall replied:
@amilcarlucas Sorry, what are you saying here?

amilcarlucas replied:
You renamed "desired" to "target" everywhere.
I assume you also want to change:

-#define PID_LABELS "TimeUS,Des,Act,Err,P,I,D,FF"
+#define PID_LABELS "TimeUS,Tar,Act,Err,P,I,D,FF"

lthall replied:
Ahh, bingo. Thanks!!

lthall commented:
Thank you everybody for your help and support to get this work in!!

IamPete1 commented:
@lthall @rmackay9 We seem to have lost the sailboat changes at some point, they defiantly here when I checked them, do you have them in a branch somewhere?

lthall commented:
Hi @IamPete1 I believe they are in APM_Control.

IamPete1 commented:
@lthall It is broken on master, that is it runs but causes the sail to be fully out all the time, there should be zero error as the the SITL doesn't model roll, so I don't think it is just a change of defaults. I thought I remembered some changes to sailboat.cpp when I had a look, possibly I miss remembered. I will see if I can fix.

lthall commented:
Ahh, ok. Sorry, I thought the changes went missing. Sailboat was one of the ones that was implemented in a strange way. I will take a look as well.

IamPete1 commented:
I thought I had seen something that had gone missing. Looks like it may just be a case of bad defaults. The old defaults seem to cause much larger response than on the old PID's. This is abit like what a saw on tracker, would you expect that?

lthall commented:
@IamPete1 it looks like your original error calculation was done incorrectly.
heel_error =  fabsf(_ahrs.roll) - desired_heel;
This will result in -1 * error. 
Convention is to calculate error = desired - measured.

I suspect I saw this when I put it in the first time and used update_error instead. This may then have been changed to update_all.

It would be great if you could continue to use update_all and fix the surrounding code because I want to remove update_error from the PID object.

I am very sorry this messed you around!!

IamPete1 commented:
no worries, quite a simple fix in the end. https://github.com/ArduPilot/ardupilot/pull/11423/commits/8a62a0df86bc11a0cf27ad0d8316d71348be7fec

My fault for the original bad implementation, thanks for the pointers on the fix, have tested this on the sailboat although admittedly on the stand rather than the water.
